docs/bounded_dynamism.md

```diff
@@ -0,0 +1,85 @@
+# StableHLO Bounded Dynamism
+
+[StableHLO](https://github.com/openxla/stablehlo) is an operation set that expresses ML computations. It has been originally bootstrapped from [the MHLO dialect](https://github.com/tensorflow/mlir-hlo#meta-hlo-dialect-mhlo), including inheriting the type and some of the ops. This RFC aims to describe the current status and future plans for bounded dynamism in StableHLO. In particular, this RFC doesn’t propose any further major changes to the current state. The only minor change proposed is allowing `i64` typed result for `get_dimension_size` op and `i64` type for `set_dimension_size` `size` operand. This is further described in P3.
+
+Bounded dynamism allows computations to encode the maximum runtime size of dynamic dimensions of tensors. This makes it possible to run particular computations on hardware that doesn’t support dynamic tensors and needs hints on the largest possible. For example, real time inference might require support for dynamic batch so that inference doesn’t need to wait for accumulation up to the static batch size. Similarly, during training the last batch in a dataset might be smaller than the max batch size if the number of examples in the dataset is not divisible by the batch size.
+
+
+# Proposal
+
+
+## (P1) Use RankedTensorType encoding field for bounds
+
+Bounds for a dynamic tensor are represented using the `TypeExtensionsAttr` using the encoding field in the `RankedTensorType`. Bounds in `TypeExtensionsAttr` is an `int64_t` array of size equal to rank of the tensor. Values corresponding to static dimensions must be `ShapedType::kDynamicSize` which is `-1`.
+
+For example, the following type represents a 2D tensor with up to `3` rows and exactly `5` columns.
+
+
+```
+tensor<?x5xf32, #stable_hlo.type_extensions<bounds = [3, -1]>>
```

Comment:
Can you motivate why static sizes get a dynamic bound and not the static known bound? Is there some advantage to this? Thinking further, why would it not be legal (even if not very useful) to have a bound that is larger than the static size?

Reply:
There is only a weak reason for this choice but regardless better to be consistent. Either, we use static bounds for static sizes or use dynamic bound but not allowing both. 

By using dynamic bound for static sizes, we know what dims are dynamic by looking at the bounds. 

If we have bound larger than the static size, this won't be consistent as there are many ways to represent the same actual information. This also makes it different to round trip with XLA:HLO that uses a bool array to signify if a particular dim is static or is a bound.

Reply:
Thanks! Added rationale in the doc.